### PR TITLE
QA-327 Actions collector

### PR DESCRIPTION
**Pull request by dgusakov**

### CODE REVIEW

File: src/adcm_pytest_plugin/objects/actions.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    action_name: str
+    parent_name: str
+    parent_type: str
+    bundle_info: str
+    expected_status: str
+    called_from: str
+
+    def to_dict(self):
+        """Convert object to dict"""
+        return self.__dict__
```
Comment (madlabman) [inline src/adcm_pytest_plugin/objects/actions.py:35]: Why not to use builtin `asdict` method?

File: src/adcm_pytest_plugin/objects/actions.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    @classmethod
+    def from_dict(cls, source_dict: dict):
+        """Recreate instance from json string"""
+        return cls(**{key: source_dict[key] for key in cls.__dict__["__annotations__"].keys()})
```
Comment (madlabman) [inline src/adcm_pytest_plugin/objects/actions.py:40]: Use `dataclasses.fields` names.

File: src/adcm_pytest_plugin/objects/actions.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    actions: List[ActionRunInfo]
+    actions_specs: List[ActionsSpec]
+
+    def make_summary(self) -> str:
+        """Make summary report in form of JSON string"""
+
+        def nested_dict():
+            """Nested dict helper"""
+            return defaultdict(nested_dict)
```
Comment (madlabman) [inline src/adcm_pytest_plugin/objects/actions.py:119]: Nice trick, but write a more descriptive doc-string.

File: src/adcm_pytest_plugin/plugin.py [hunk 4]
```diff
@@ -234,10 +252,68 @@ def pytest_runtest_makereport(item, call):
     setattr(item, "rep_" + rep.when, rep)
 
 
-def pytest_sessionfinish():
+def _get_actions_dir(config):
+    return (
+        os.path.join(os.getcwd(), config.option.actions_report_dir)
```
Comment (madlabman) [inline src/adcm_pytest_plugin/plugin.py:257]: What if to use `Path` type for value of `--actions-report-dir`?

**Review by madlabman — CHANGES_REQUESTED** [review]

File: src/adcm_pytest_plugin/plugin.py [hunk 5]
```diff
@@ -40,6 +44,13 @@ def some_func():
     """
     global options  # pylint: disable=global-statement,invalid-name,global-variable-not-assigned
     options.__dict__.update(config.option.__dict__)
+    if config.option.actions_report_dir:
+        pytest.action_run_storage = []
+        pytest.actions_spec_storage = {}
+        try:
+            shutil.rmtree(_get_actions_dir(config))
+        except FileNotFoundError:
+            pass
```
Comment (vgorkavenko) [inline src/adcm_pytest_plugin/plugin.py:53]: If we run tests with `xdist` this can break our report

Comment (dgusakov) [reply src/adcm_pytest_plugin/plugin.py:53]: How?

Comment (vgorkavenko) [reply src/adcm_pytest_plugin/plugin.py:53]: Each test will run `rmtree`

Comment (madlabman) [reply src/adcm_pytest_plugin/plugin.py:53]: Hook will be called only once, as I understand.

File: src/adcm_pytest_plugin/steps/actions.py [hunk 6]
```diff
@@ -312,7 +314,14 @@ def _run_action_and_assert_result(
...
         obj.reread()
         try:
-            task = obj.action(name=action_name).run(**kwargs)
+            action = obj.action(name=action_name)
+            if getattr(options, "actions_report_dir", None):
+                pytest.action_run_storage.append(
+                    ActionRunInfo.from_action(action=action, expected_status=expected_status)
+                )
+                actions_spec = ActionsSpec.from_action(action=action)
+                pytest.actions_spec_storage[actions_spec.uniq_id] = actions_spec
```
Comment (Sealwing) [inline src/adcm_pytest_plugin/steps/actions.py:323]: Maybe move it to helper function with an early exit to not make an accent of a variable feature.
Like that:
``` python
def _add_action_to_report(options, action, expected_status): 
    if not getattr(options, "actions_report_dir", None):
        return
    pytest.action_run_storage.append(
        ActionRunInfo.from_action(action=action, expected_status=expected_status)
    )
    actions_spec = ActionsSpec.from_action(action=action)
    pytest.actions_spec_storage[actions_spec.uniq_id] = actions_spec
```
of even leave options check as is and move only report logic

**Review by Sealwing — CHANGES_REQUESTED** [review]

File: src/adcm_pytest_plugin/plugin.py [hunk 7]
```diff
@@ -234,10 +252,68 @@ def pytest_runtest_makereport(item, call):
...
+            )
+        del pytest.actions_spec_storage
+
+
+def pytest_unconfigure(config: Config):
+    """Create actions call report"""
+    if not os.getenv("PYTEST_XDIST_WORKER") and config.option.actions_report_dir:
+        summary_file = "summary.json"
+        common_actions_call_list = []
+        common_actions_spec = {}
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply src/adcm_pytest_plugin/plugin.py:298]: Since it is 2 different objects

[the remark this answers is not in the document]
Comment (Sealwing) [reply src/adcm_pytest_plugin/plugin.py:298]: I mean why "common" actions_call_list? I just don't get: common between what? :D

File: src/adcm_pytest_plugin/plugin.py (hunk 5 above)
Comment (dgusakov) [reply src/adcm_pytest_plugin/plugin.py:53]: The hook will be called ones per thread

Comment (vgorkavenko) [reply src/adcm_pytest_plugin/plugin.py:53]: Hook will be called for each thread, but this logic actually does not break the report

File: src/adcm_pytest_plugin/plugin.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dgusakov) [reply src/adcm_pytest_plugin/plugin.py:298]: Common between tests in one thread

[the remark this answers is not in the document]
Comment (Sealwing) [reply src/adcm_pytest_plugin/plugin.py:298]: Got it, good

File: src/adcm_pytest_plugin/objects/actions.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+
+@dataclass
+class ActionsSpec:
+    """Info about actions from prototype"""
+
+    actions: List[str]
+    parent_name: str
+    parent_type: str
+    bundle_info: str
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply src/adcm_pytest_plugin/objects/actions.py:65]: This naming comes from the object above. Actions+Their parent looks ok for me.

File: src/adcm_pytest_plugin/plugin.py [hunk 9]
```diff
@@ -234,10 +252,68 @@ def pytest_runtest_makereport(item, call):
...
+        for filename in os.listdir(actions_report_dir):
+            if "run" in filename:
+                with open(os.path.join(actions_report_dir, filename), "r", encoding="utf-8") as file:
+                    common_actions_call_list.extend([ActionRunInfo.from_dict(obj) for obj in json.loads(file.read())])
+            if "spec" in filename:
+                with open(os.path.join(actions_report_dir, filename), "r", encoding="utf-8") as file:
+                    for uniq_id, actions_spec_dict in json.loads(file.read()).items():
+                        actions_spec = ActionsSpec.from_dict(actions_spec_dict)
+                        common_actions_spec[uniq_id] = actions_spec
+            os.remove(os.path.join(actions_report_dir, filename))
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply src/adcm_pytest_plugin/plugin.py:309]: Changed

File: src/adcm_pytest_plugin/steps/actions.py (hunk 6 above)
Comment (dgusakov) [reply src/adcm_pytest_plugin/steps/actions.py:323]: Done

File: src/adcm_pytest_plugin/objects/actions.py (hunk 3 above)
Comment (dgusakov) [reply src/adcm_pytest_plugin/objects/actions.py:119]: Done

File: src/adcm_pytest_plugin/objects/actions.py (hunk 2 above)
Comment (dgusakov) [reply src/adcm_pytest_plugin/objects/actions.py:40]: Done

File: src/adcm_pytest_plugin/objects/actions.py (hunk 1 above)
Comment (dgusakov) [reply src/adcm_pytest_plugin/objects/actions.py:35]: Done

File: src/adcm_pytest_plugin/objects/actions.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Sealwing) [reply src/adcm_pytest_plugin/objects/actions.py:65]: I just see "parent" as an ancestor, but here it is an owner. But if it's okay for others, I think it is clear enough

File: src/adcm_pytest_plugin/plugin.py (hunk 4 above)
Comment (dgusakov) [reply src/adcm_pytest_plugin/plugin.py]: Done

File: src/adcm_pytest_plugin/plugin.py [hunk 10]
```diff
@@ -234,10 +255,64 @@ def pytest_runtest_makereport(item, call):
...
     Clear custom env variables at the end of all tests
+    Create files with raw actions call data
     """
     for var in dict(os.environ):
         if "rep_" in var:
             del os.environ[var]
+
+    if session.config.option.actions_report_dir:
+        actions_report_dir = _get_actions_dir(session.config)
+        Path(actions_report_dir).mkdir(parents=True, exist_ok=True)
```
Comment (madlabman) [inline src/adcm_pytest_plugin/plugin.py:273]: It's  a `Path` instance already.

Comment (dgusakov) [reply src/adcm_pytest_plugin/plugin.py]: Done

**Review by Sealwing — APPROVED** [review]

**Review by madlabman — APPROVED** [review]

**Pull request merged**